commit 0d5b5a0f32dfc9be3521803aba53e856759f0250
Author: Chuck Lever <chuck.lever@oracle.com>
Date:   Sun Jan 15 12:22:56 2023 -0500

    SUNRPC: Add RFC 8009 encryption and decryption functions
    
    RFC 8009 enctypes use different crypt formulae than previous
    Kerberos 5 encryption types. Section 1 of RFC 8009 explains the
    reason for this change:
    
    > The new types conform to the framework specified in [RFC3961],
    > but do not use the simplified profile, as the simplified profile
    > is not compliant with modern cryptographic best practices such as
    > calculating Message Authentication Codes (MACs) over ciphertext
    > rather than plaintext.
    
    Add new .encrypt and .decrypt functions to handle this variation.
    
    The new approach described above is referred to as Encrypt-then-MAC
    (or EtM). Hence the names of the new functions added here are
    prefixed with "krb5_etm_".
    
    A critical second difference with previous crypt formulae is that
    the cipher state is included in the computed HMAC. Note however that
    for RPCSEC, the initial cipher state is easy to compute on both
    initiator and acceptor because it is always all zeroes.
    
    Tested-by: Scott Mayhew <smayhew@redhat.com>
    Reviewed-by: Simo Sorce <simo@redhat.com>
    Signed-off-by: Chuck Lever <chuck.lever@oracle.com>

diff --git a/net/sunrpc/auth_gss/gss_krb5_mech.c b/net/sunrpc/auth_gss/gss_krb5_mech.c
index b55897cac459..7fe39e47c84f 100644
--- a/net/sunrpc/auth_gss/gss_krb5_mech.c
+++ b/net/sunrpc/auth_gss/gss_krb5_mech.c
@@ -167,6 +167,8 @@ static const struct gss_krb5_enctype supported_gss_krb5_enctypes[] = {
 
 		.import_ctx	= gss_krb5_import_ctx_v2,
 		.derive_key	= krb5_kdf_hmac_sha2,
+		.encrypt	= krb5_etm_encrypt,
+		.decrypt	= krb5_etm_decrypt,
 
 		.get_mic	= gss_krb5_get_mic_v2,
 		.verify_mic	= gss_krb5_verify_mic_v2,
@@ -192,6 +194,8 @@ static const struct gss_krb5_enctype supported_gss_krb5_enctypes[] = {
 
 		.import_ctx	= gss_krb5_import_ctx_v2,
 		.derive_key	= krb5_kdf_hmac_sha2,
+		.encrypt	= krb5_etm_encrypt,
+		.decrypt	= krb5_etm_decrypt,
 
 		.get_mic	= gss_krb5_get_mic_v2,
 		.verify_mic	= gss_krb5_verify_mic_v2,